newIDE/app/src/VariablesList/index.js

```diff
@@ -254,43 +252,78 @@ export default class VariablesList extends React.Component<Props, State> {
     );
   }
 
-  render() {
+  onAddVariable = (variableName = false) => {
     const { variablesContainer } = this.props;
+    const variable = new gd.Variable();
+    variable.setString('');
+
+    const name = variableName
+      ? variableName
+      : newNameGenerator('Variable', name => variablesContainer.has(name));
+    variablesContainer.insert(name, variable, variablesContainer.count());
+    return variable;
+  };
+
+  render() {
+    const { variablesContainer, objectVariables } = this.props;
     if (!variablesContainer) return null;
 
+    /// map all instance variables with the same name as object variables
+    const objVariablesTree = objectVariables
+      ? mapFor(0, objectVariables.count(), index => {
+          const name = objectVariables.getNameAt(index);
+          if (!variablesContainer.has(name)) {
+            this.onAddVariable(name);
```

Comment:
I'm still uncomfortable doing this. Doing changes in a render method is usually a sign that something is wrong.
This means here that as soon as someone *display* the instance variables of an instance, then automatically all object variables get added to the list of instance variables.

Worse, by doing this you actually overwrite all object variables by instance variables that have an empty value... so if I'm not mistaken you end up with an instance with all variables set to an empty value.

You should only add a variable to the `variablesContainer` when the user try to change the value of a variable that used to be a variable from the `objectVariablesContainer`.

This means that you have to take a slightly different approach: you have to iterate on instance AND object variables, and for each object variables, only display it if it's NOT in the instance variables container.

Take a bit of time to get familiar with the code of this file, `VariablesList/index.js`. What you want to do is not an easy task, take your time and try to see how the iteration is made on existing variables (actually, it's a "tree traversal" because variables can be structures).
Then, see how to display "parent variables" (what we are calling object variables should be called parent variables in fact, they are existing and "inherited" by instance variables).

Reply:
I already iterate on instance and object variables here, two mapFors that create the trees, which are then fed to be  rendered. What I'm not sure about is how do I get it to automatically add instance variables of object variables

Reply:
Basically, think that you are totally forbidden to call addVariable in render.
When would you then call addVariable? What is this place in the code?

Reply:
It is now moved to being called on mounting

Reply:
I think there have been misunderstanding - sorry about that. The right answer was: "call addVariable when an object variable is edited". Because it's easy to confuse things, object variables here should be renamed "parent variables".